### PR TITLE
feat: add standalone mode in java sdk

### LINKED ISSUE #620: add standalone mode in java sdk
requirement：
openmldb server has standalone cluster, but this standalone mode is missing in java sdk. 
maybe provide standalone mode in client  is better

### LINKED ISSUE #1523: PRE_AGG_META table can't create in standalone mode
image

### PR DESCRIPTION
**Pull request by keyu813**
close #620 
`onebox/start_onebox.sh` support standalone mode
close #1523

### CODE REVIEW

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java [hunk 1]
```diff
@@ -58,24 +44,34 @@
...
+            this.sqlRouter = sql_router_sdk.NewStandaloneSQLRouter(sqlOpt);
+        }
         if (sqlRouter == null) {
             throw new SqlException("fail to create sql executor");
         }
     }
 
-    public SqlClusterExecutor(SdkOption option) throws SqlException {
-        this(option, "sql_jsdk");
+    public SqlClusterExecutor(SdkOption option, boolean isClusterMode) throws SqlException {
```
Comment (dl239) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java]: move mode to `SdkOption`

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java [hunk 2]
```diff
@@ -22,6 +22,10 @@
     private long sessionTimeout = 10000;
     private Boolean enableDebug = false;
     private long requestTimeout = 60000;
+    private String host;
+    private long port;
+
+    private boolean isClusterMode;
```
Comment (dl239) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java]: add default value `true`, so the exist case need not update

**Comment by dl239** [comment]
add standalone test. start standalone openmldb in `onebox/start_onebox.sh`

**Comment by keyu813** [comment]
> add standalone test. start standalone openmldb in `onebox/start_onebox.sh`

ok, should I add a StandaloneSmokeTest in `/test` ? or just update `SQLRouterSmokeTest.java` ?

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/StandaloneRouterSmokeTest.java [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+    static {
+        try {
+            SdkOption option = new SdkOption();
+            option.setClusterMode(false);
+            option.setHost(TestConfig.HOST);
+            option.setPort(TestConfig.PORT);
+            option.setSessionTimeout(200000);
+            router = new SqlClusterExecutor(option);
+            java.sql.Statement state = router.getStatement();
+            state.execute("SET @@execute_mode='online';");
```
Comment (dl239) [inline java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/StandaloneRouterSmokeTest.java]: standalone need not set online mode

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/StandaloneRouterSmokeTest.java [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+            state.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Test
+    public void testSmoke() {
+        try {
```
Comment (dl239) [inline java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/StandaloneRouterSmokeTest.java]: if the case is same as Cluster smoketest, it's better to use `DataProvider`

**Review by dl239 — APPROVED** [review]

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java [hunk 5]
```diff
@@ -16,20 +16,7 @@
...
-import com._4paradigm.openmldb.SQLInsertRows;
-import com._4paradigm.openmldb.SQLRequestRow;
-import com._4paradigm.openmldb.SQLRouter;
-import com._4paradigm.openmldb.SQLRouterOptions;
-import com._4paradigm.openmldb.Status;
-import com._4paradigm.openmldb.TableColumnDescPair;
-import com._4paradigm.openmldb.TableColumnDescPairVector;
-import com._4paradigm.openmldb.TableReader;
-import com._4paradigm.openmldb.VectorString;
+import com._4paradigm.openmldb.*;
```
Comment (vagetablechicken) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java]: Do not change it to’ import *’

**Review by vagetablechicken — APPROVED** [review]

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/TestConfig.java [hunk 6]
```diff
@@ -21,4 +21,6 @@
     public static String ZK_PATH="/onebox";
 
 
+    public static String HOST="127.0.0.1";
```
Comment (aceforeverd) [inline java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/TestConfig.java:25]: better comment a few lines

File: onebox/start_onebox.sh [hunk 7]
```diff
@@ -20,76 +20,123 @@ cd "$(dirname "$0")/../"
...
+
+        # start ns3
+        ./build/bin/openmldb --endpoint=${NS3} --role=nameserver \
+                           --tablet_offline_check_interval=1\
+                           --tablet_heartbeat_timeout=1\
+                           --zk_cluster=${ZK_CLUSTER}\
+                           --zk_root_path=/onebox > ns3.log 2>&1 &
+        sleep 2
+        echo "cluster start ok"
+        exit 0
```
Comment (aceforeverd) [inline onebox/start_onebox.sh]: do not exit in function

File: onebox/start_onebox.sh [hunk 8]
```diff
@@ -20,76 +20,123 @@ cd "$(dirname "$0")/../"
...
+        sleep 2
+
+        # start ns
+        ./build/bin/openmldb --endpoint=${NS} --role=nameserver \
+                           --tablet=${TABLET} \
+                           --tablet_offline_check_interval=1\
+                           --tablet_heartbeat_timeout=1 > ns.log 2>&1 &
+        sleep 2
+        echo "standalone start ok"
+        exit 0
```
Comment (aceforeverd) [inline onebox/start_onebox.sh]: same here

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java [hunk 9]
```diff
@@ -61,14 +61,25 @@
...
-        sqlOpt.delete();
+        if (option.isClusterMode()) {
+            SQLRouterOptions sqlOpt = new SQLRouterOptions();
+            sqlOpt.setSession_timeout(option.getSessionTimeout());
+            sqlOpt.setZk_cluster(option.getZkCluster());
+            sqlOpt.setZk_path(option.getZkPath());
+            sqlOpt.setEnable_debug(option.getEnableDebug());
+            sqlOpt.setRequest_timeout(option.getRequestTimeout());
+            this.sqlRouter = sql_router_sdk.NewClusterSQLRouter(sqlOpt);
+            sqlOpt.delete();
```
Comment (aceforeverd) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java:73]: don't the GC handle it automatically ? @vagetablechicken

File: .github/workflows/sdk.yml [hunk 10]
```diff
@@ -91,7 +91,7 @@ jobs:
         run: |
           sh steps/ut_zookeeper.sh start
           sh steps/download_openmldb_spark.sh $SPARK_HOME
-          cd onebox && sh start_onebox.sh && cd - || exit
+          cd onebox && sh start_onebox.sh && sh start_onebox.sh standalone && cd - || exit
```
Comment (aceforeverd) [inline .github/workflows/sdk.yml:94]: the same should be in coverage.yml

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java [hunk 11]
```diff
@@ -62,4 +66,28 @@ public long getRequestTimeout() {
     public void setRequestTimeout(long requestTimeout) {
         this.requestTimeout = requestTimeout;
     }
+
+    public String getHost() {
```
Comment (aceforeverd) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java]: let's use Lombok ?

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java [hunk 12]
```diff
@@ -41,25 +42,38 @@
...
 
     static {
         try {
             SdkOption option = new SdkOption();
             option.setZkPath(TestConfig.ZK_PATH);
             option.setZkCluster(TestConfig.ZK_CLUSTER);
             option.setSessionTimeout(200000);
-            router = new SqlClusterExecutor(option);
-            java.sql.Statement state = router.getStatement();
+            clusterRouter = new SqlClusterExecutor(option);
```
Comment (aceforeverd) [inline java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java]: must call `Init()` for both `clusterRouter` and `standaloneRouter` before use

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java [hunk 13]
```diff
@@ -62,4 +66,28 @@ public long getRequestTimeout() {
...
+
+    public void setPort(long port) {
+        this.port = port;
+    }
+
+    public boolean isClusterMode() {
+        return isClusterMode;
+    }
+
+    public void setClusterMode(boolean clusterMode) {
```
Comment (aceforeverd) [inline java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java]: where the option is set ?

Comment (keyu813) [reply java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java:90]: before use SqlClusterExecutor we should set option, see SQLRouterSmokeTest

Comment (keyu813) [reply java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/SdkOption.java]: > where the option is set ?

before use SqlClusterExecutor we should set option, see SQLRouterSmokeTest

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java (hunk 12 above)
Comment (keyu813) [reply java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java]: > must call `Init()` for both `clusterRouter` and `standaloneRouter` before use

which init()? SqlClusterExecutor seems have no `Init()` method

File: java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java (hunk 9 above)
Comment (vagetablechicken) [reply java/openmldb-jdbc/src/main/java/com/_4paradigm/openmldb/sdk/impl/SqlClusterExecutor.java:73]: If `swigCMemOwn==true`, it'll gc. But we've found CMem cost too much or memory leak before(maybe gc is not immediate), so in java sdk, we manually delete CMem in every place.
Just leave it, until we can remove all manul deletion.

File: java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java (hunk 12 above)
Comment (vagetablechicken) [reply java/openmldb-jdbc/src/test/java/com/_4paradigm/openmldb/jdbc/SQLRouterSmokeTest.java]: `router->Init()` is in `SqlClusterExecutor`.  It's better to change the name.  clusterRouter -> clusterExecutor or clusterClient

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**